os/fs/smartfs/smartfs_utils.c

```diff
@@ -1268,6 +1295,241 @@ int smartfs_createentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector,
+
+	/* Scan through the entire parent sector chain to find an invalid entry */
+	while (psector != SMARTFS_ERASEDSTATE_16BIT) {
+		/* Read the next sector */
+
+		readwrite.logsector = psector;
+		readwrite.count = fs->fs_llformat.availbytes;
+		readwrite.offset = 0;
+		readwrite.buffer = (uint8_t *)fs->fs_rwbuffer;
+		ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+		if (ret < 0) {
+			goto errout;
+		}
+
+		/* Get the next chained sector */
+		chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+		nextsector = SMARTFS_NEXTSECTOR(chainheader);
+
+		/* Search for an invalid entry in this sector */
+		offset = sizeof(struct smartfs_chain_header_s);
+		entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
+
+		while ((offset + entrysize) < readwrite.count) {
+			/* Check if this entry is invalid */
+
+			if (entry->flags != SMARTFS_ERASEDSTATE_16BIT && !(ENTRY_VALID(entry))) {
+				invalidentry->dsector = psector;
+				invalidentry->doffset = offset;
+				ret = OK;
+				goto errout;
+			}
+
+			/* If we find an empty entry in the process and no invalidated entry is found
+			   in the end, we can return the empty entry we had found.
+			 */
+			if (saved_doffset == -1 && IS_EMPTY_ENTRY(entry)) {
+				saved_psector = psector;
+				saved_doffset = offset;
+			}
+
+			/* Not and invalid entry, check the next one */
+			offset += entrysize;
+			entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
+		}
+		psector = nextsector;
+	}
+
+errout:
+	if (ret == OK) {
+		RESET_ENTRY_FLAGS(invalidentry);
+	} else {
+		if (saved_doffset != -1) {
+			invalidentry->dsector = saved_psector;
+			invalidentry->doffset = saved_offset;
+			ret = OK;
+		}
+	}
+	return ret;
+}
+
+/****************************************************************************
+ * Name: smartfs_replaceentry
+ *
+ * Description: Takes care of the operation of renaming an existing file
+ *              in a particular directory to a new name in the same dircetory
+ *              while minimizing the need to allocate new sectors for
+ *              writing the changed entry to MTD flash.
+ *
+ ****************************************************************************/
+
+int smartfs_replaceentry(struct smartfs_mountpt_s *fs, struct smartfs_entry_s oldentry, const char *filename, uint16_t type, mode_t mode, struct smartfs_entry_s *direntry, FAR struct smartfs_ofile_s *sf)
+{
+	int ret;
+	struct smart_read_write_s readwrite;
+	uint16_t psector;
+	uint16_t offset;
+	uint16_t nextsector;
+	uint16_t entrysize;
+	struct smartfs_chain_header_s *chainheader;
+	uint8_t chainheader_type;
+	struct smartfs_entry_header_s *entry;
+
+	entrysize = sizeof(struct smartfs_entry_header_s) + fs->fs_llformat.namesize;
+	psector = oldentry.dsector;
+	offset = oldentry.doffset;
+
+#ifdef CONFIG_SMARTFS_ALIGNED_ACCESS
+	if (smartfs_rdle16(&oldentry.flags) == SMARTFS_ERASEDSTATE_16BIT) {
+#else
+	if (oldentry.flags == SMARTFS_ERASEDSTATE_16BIT) {
+#endif
```

Comment:
Why you don't you make MACRO for this? like a wrle16?

Reply:
I dont think that is necessary.
Also this kind of check for flags in ERASEDSTATE is performed only once.